### PR TITLE
Adding different module resolution strategies

### PR DESCRIPTION
**Pull request by vladima**

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -36,11 +36,169 @@ namespace ts {
...
+            return false;
+        }
+
+        // module name starts with './' or '../'
+        if (nameStartsWithDotSlashOrDotDotSlash(moduleName)) {
+            return false;
+        }
+        
+        // module name has one of supported extesions
+        for(let ext of supportedExtensions ) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: I do not think we need this since SystemJS allowes it

File: src/compiler/program.ts [hunk 2]
```diff
@@ -36,11 +36,169 @@ namespace ts {
...
+        
+        let normalizedModuleName = normalizeSlashes(moduleName);         
+        let basePart = useBaseUrl(moduleName) ? baseUrl : getDirectoryPath(containingFile);        
+        let candidate = normalizePath(combinePaths(basePart, moduleName));
+        
+        let failedLookupLocations: string[] = [];
+        
+        let hasSupportedExtension = forEach(supportedExtensions, ext => fileExtensionIs(candidate, ext));
+
+        if (hasSupportedExtension) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: let's start with not allowing file extensions, we can relax that later on.

File: src/compiler/program.ts [hunk 3]
```diff
@@ -36,11 +36,169 @@ namespace ts {
...
+            // module name already has extension - use it as is
+            let result = tryLoadFile(candidate);
+            if (result) {
+                return result;
+            }            
+        }
+        else {
+            // module name does not have extension - try every supported extension
+            for(let ext of supportedExtensions) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: nit: forEach(supportedExtensions, ext => tryLoadFile(candidate + ext));

File: src/compiler/program.ts [hunk 4]
```diff
@@ -36,11 +36,169 @@ namespace ts {
     }
     
     export function resolveModuleName(moduleName: string, containingFile: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost): ResolvedModule {
-        // TODO: use different resolution strategy based on compiler options
-        return legacyNameResolver(moduleName, containingFile, compilerOptions, host);
+        switch(compilerOptions.moduleResolution) {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Space after switch

File: src/compiler/program.ts (hunk 3 above)
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Space after `for`

File: src/compiler/program.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Space after `for`

File: src/compiler/program.ts [hunk 5]
```diff
@@ -36,11 +36,169 @@ namespace ts {
...
+                }
+                
+                result = loadNodeModuleFromDirectory(candidate, failedLookupLocations, host);
+                if (result) {
+                    return { resolvedFileName: result, failedLookupLocations };
+                }
+            }
+            
+            let parentPath = getDirectoryPath(directory);
+            if (parentPath === directory) {
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Add a comment to explain what's going on here.

File: src/compiler/program.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/program.ts]: ok

File: src/compiler/program.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/program.ts]: ok

File: src/compiler/program.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/program.ts]: ok

File: tests/cases/unittests/moduleResolution.ts [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        function isDirectory(fse: Directory | File): boolean {
+            return (<Directory>fse).children !== undefined;
+        }
+        
+        function makeFS(files: File[]): Directory {
+            // create root
+            let {dir} = splitPath(files[0].name);
+            let root: Directory = { name: dir, children: {} };
+            
+            for(let f of files) {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/moduleResolution.ts]: `for` space

**Comment by vilic** [comment]
!!! Fantastic, finally got here. Thanks!

File: src/compiler/types.ts [hunk 7]
```diff
@@ -2008,7 +2008,13 @@ namespace ts {
         Error,
         Message,
     }
-
+    
+    export const enum ModuleResolutionKind {
+        Legacy  = 1,
```
Comment (yuit) [inline src/compiler/types.ts]: What is Legacy? Could you add comment?

File: src/compiler/program.ts [hunk 8]
```diff
@@ -36,11 +36,136 @@ namespace ts {
...
+            }
+            
+            directory = parentPath;
+        }
+        
+        return { resolvedFileName: undefined, failedLookupLocations };
+    }
+    
+    export function baseUrlModuleNameResolver(moduleName: string, containingFile: string, baseUrl: string, host: ModuleResolutionHost): ResolvedModule {
+        Debug.assert(baseUrl !== undefined);
```
Comment (yuit) [inline src/compiler/program.ts]: Should there be a check and give out compilerOptions error if the baseUrl is not defined?

File: src/compiler/program.ts [hunk 9]
```diff
@@ -36,11 +36,136 @@ namespace ts {
...
+        }
+    }
+    
+    export function nodeModuleNameResolver(moduleName: string, containingFile: string, host: ModuleResolutionHost): ResolvedModule {
+        let containingDirectory = getDirectoryPath(containingFile);        
+
+        if (getRootLength(moduleName) !== 0 || nameStartsWithDotSlashOrDotDotSlash(moduleName)) {
+            let failedLookupLocations: string[] = [];
+            let candidate = normalizePath(combinePaths(containingDirectory, moduleName));
+            let result = loadNodeModuleFromFile(candidate, failedLookupLocations, host);
```
Comment (yuit) [inline src/compiler/program.ts]: use ```let resolvedFileName``` instead

File: src/compiler/program.ts [hunk 10]
```diff
@@ -36,11 +36,136 @@ namespace ts {
...
+            case ModuleResolutionKind.BaseUrl: return baseUrlModuleNameResolver(moduleName, containingFile, compilerOptions.baseUrl, host);
+            default: return legacyNameResolver(moduleName, containingFile, compilerOptions, host);
+        }
+    }
+    
+    export function nodeModuleNameResolver(moduleName: string, containingFile: string, host: ModuleResolutionHost): ResolvedModule {
+        let containingDirectory = getDirectoryPath(containingFile);        
+
+        if (getRootLength(moduleName) !== 0 || nameStartsWithDotSlashOrDotDotSlash(moduleName)) {
+            let failedLookupLocations: string[] = [];
```
Comment (yuit) [inline src/compiler/program.ts]: Is this used to give out any error or it is mainly use for testing?

File: tests/cases/unittests/moduleResolution.ts (hunk 6 above)
Comment (vladima) [inline tests/cases/unittests/moduleResolution.ts]: ok

File: src/compiler/program.ts (hunk 10 above)
Comment (vladima) [inline src/compiler/program.ts]: this is used as the part of output

File: src/compiler/program.ts (hunk 9 above)
Comment (vladima) [inline src/compiler/program.ts]: ok

File: src/compiler/program.ts (hunk 8 above)
Comment (vladima) [inline src/compiler/program.ts]: this check will be in `verifyCompilerOptions` - it is not there yet

**Comment by alexeagle** [comment]
+cc @mprobst

File: src/harness/harness.ts [hunk 11]
```diff
@@ -1025,6 +1025,9 @@ module Harness {
                                     options.module = ts.ModuleKind.UMD;
                                 } else if (setting.value.toLowerCase() === "commonjs") {
                                     options.module = ts.ModuleKind.CommonJS;
+                                    if (options.moduleResolution === undefined) {
+                                        options.moduleResolution = ts.ModuleResolutionKind.Classic;
```
Comment (mhegazy) [inline src/harness/harness.ts]: can you add a comment

**Comment by mhegazy** [comment]
:+1:

**Comment by mihailik** [comment]
Is there a description of what this is doing and how to use it please?

**Comment by vilic** [comment]
@mihailik You may specify "moduleResolution" in compilerOptions as "node", then it will try to load declaration file under related node module. For me it seems that either `index.d.ts` or specified "typings" field in package.json works, not typings related to "main" field.

**Comment by mhegazy** [comment]
one thing to add, `--m commonjs` means your `moduleResolution` is node, so if this is what you are using, you get the new behavior automatically.

let us know if you run into issues with the new resolution logic.

**Pull request merged**